### ISSUE TITLE
js-native - Error: Can't dispatch XXXXXXX in JS_NATIVES table

### ISSUE THREAD

**Issue by BrianOtto** [metaeducation/ren-c#952]
The following issue happens at http://hostilefork.com/media/shared/replpad-js/ and on my local builds in both Windows and Linux. The `js-native` extension is not working. Try running these commands ...

`test: js-native [] "console.log(123);"`
`test`

As far as I understand it, this should write "123" to the console. I have tested this with different variations of the JavaScript code, and none of them work. I can provide more examples if you need. It will always return the following error (the number changes depending on what JS code you use).

`Error: Can't dispatch 6093048 in JS_NATIVES table`

I have done some initial debugging, and it appears like it might be a scoping issue. If you put a `console.log(RL_JS_NATIVES)` at the beginning of `reb.RegisterId_internal` and `reb.RunNative_internal`, in `libr3.js`, you will see the `RL_JS_NATIVES` objects contain different data between the two functions. It appears like `reb.RegisterId_internal` is registering it's IDs in a copy of the object and that's why `reb.RunNative_internal` can't find/dispatch the JS code.

The `RL_JS_NATIVES` object is global, and so I don't understand why this is happening. Could it be something to do with the threading that is going on in Emscripten? Maybe those functions run in different threads and are using different copies of the object?

I am going to keep debugging the issue, but I was hoping someone might have insight into what is going on here. To be honest, the Emscripten code is pretty complex, and I don't completely, understand it all.

I believe this is important to get working because the ability to execute JavaScript could be a killer feature for replpad-js. This would give you the ability to programmatically build HTML UIs in Ren-C. Imagine a "vid" like DSL for creating an online UI. I am, it would be awesome :)

**Comment by hostilefork** [comment]
> I believe this is important to get working because the ability to execute JavaScript could be a killer feature for replpad-js. This would give you the ability to programmatically build HTML UIs in Ren-C. Imagine a "vid" like DSL for creating an online UI. I am, it would be awesome :)

Yes, this is exactly the aim.  But clawing along one step at a time to get there.  I'm hoping in the very near term (this week-ish) to start improving this proof-of-concept to more stably be able to do an accept/reject on a rebPromise().  The Rebol code you ask the promise to run is done by the worker...and if the GUI happens to come along and call rebHalt() it can interrupt that, causing the promise to reject.  As you can tell this is non-trivial to write, and to debug!

But what you are trying is very much supposed to work.

> Could it be something to do with the threading that is going on in Emscripten? 

That's very likely.  The WASM threads are new (first version of this used Emterpreter)...but they change everything.  That table is how JS gets its hands on the function, and it's a JavaScript object by definition. 
 So it can't live in the SharedArrayBuffer heap that both threads see.

I'd be thrilled if this was a lens through which you came up to speed on what's going on and could fix it!  That'd be amazing.  But don't feel like you have to.  It's a bug, and I'd certainly be fixing it if you weren't.

**Comment by BrianOtto** [comment]
I was able to come up with a solution, but you're not going to like how I implemented it :grimacing: :laughing:
I need to clean up the code a bit, but I'll commit it to a branch tomorrow.

In the mean time, you can play around with it here.
http://rebol.brianotto.com/replpad-js/

Here's a fun one to try ...
‌`test: js-native [] "document.querySelector('.note').style.background = 'red'"`
`test`

**Comment by hostilefork** [comment]
Woo!

If you hang around as a partner in crime, this Beta/One thing might happen after all!  :-)

**Comment by BrianOtto** [comment]
I have committed my changes to a new [issue-952](https://github.com/metaeducation/ren-c/tree/issue-952) branch. You can see a DIFF by going [here](https://github.com/metaeducation/ren-c/compare/master...metaeducation:issue-952?expand=1).

Let me explain what I've looked into and why I've done things this way ...

I spent a lot of time yesterday trying to implement [WebAssembly.Global()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Global) so that all worker threads could share access to the same RL_JS_NATIVES object, but nothing was working. This appears to be the "proper" solution to use, but there is very few examples of using it in the real world, and I just couldn't get it to work.

So, I started researching alternative solutions and realized you could send messages to worker threads by using the built in postMessage() function. This is exactly what I needed, as this would allow me to send a copy of the RL_JS_NATIVES object to other worked threads any time it was updated. However, I ran into an issue when I first tried it, due to the limitations of the [structured clone algorithm](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm) that Firefox and Chrome uses to serialize/transfer data.

In short, it doesn't allow me to send functions. I found a way around it though ... :wink:

Instead of storing the function in RL_JS_NATIVES, I store the string representation of that function instead (i.e.  fn.toString()). When it's time to run the function I then use \**cough, cough*\* ~evil~ eval to run it and get a result. :smiling_imp:

Ha, so this appears to be working. My main worry is that we may run into some kind of function that .toString() doesn't handle properly... but so far it's been looking good. The code that emmc generates also appears to use evals for things too, and so maybe this solution is acceptable.

What are your thoughts?

**Comment by hostilefork** [comment]
Good to see your knowledge runs deep!!  :-)

> What are your thoughts?

One thing while trying to answer you is knowing that I really do have to go back and look at this all from first principles--given what I know now--to try and explain from the ground-up exactly how this should work.  For now I'll just wing it (this code is a lot of "winging it", so far).

In the abstract, it seems like it could be worthy to make it possible to dispatch JavaScript natives off of either thread.  Clearly there's a technical pain point to doing that.  You've shown an option for sync'ing which is worth bearing in mind that it's there.

However--I think my natural bias *for the near term* is to constrain it so you can't...that you're always registering and dispatching JS natives on the same thread.  One reason to favor a simple-and-dumb answer is trying to facilitate the emterpreter solution which can run in browsers with no shared memory and no wasm threads.

It's terrible and slow to do so.  But when we think about someone clicking a link, to mess around with the web tutorial, you don't necessarily get a second chance with that person.  So I'd rather detect the browser's abilities, pull the wasm/pthread version if it can handle it, or pull the emterpreter version if not.  Then give you a message in the REPL about "hey, y'know, you could make this 30x faster if you just fiddle and reboot your browser!  do it when that's convenient." with showing them the right place to do that setting in whatever browser.

Hence mechanically, having the solution more-or-less pretend it has a single thread, probably helps write the most compatible code for the near term.  There's a kind-of-sucky property this has which you might have noticed...which is that when you resolve a promise you can't use rebXxx() APIs to build the result...because the promise's stack is suspended.  And if we have the only-run-code-on-worker mandate, that brings an additional curveball that even in the WASM build where you can call rebXxx() APIs that aren't rebPromise(), you can't put JS-NATIVE calls in that code even if they don't await.  :-/

Anyway, you're basically right-up-to-the-moment with this.  So soon (if not already) you can probably make decisions on it better than I can.

**Comment by hostilefork** [comment]
Okay, I have put my JavaScript hat back on and am re-reading this stuff and getting in the right head-space to develop on it again.

Aaaaand...I said something nearly completely wrong:

> The table should only exist on the worker thread, and the main GUI thread should not run JS natives...ever...unless it is an emscripten build and it has no choice.

The part I was right about is that it's only ever one thread--I just got wrong which one it was!  It's the *GUI* thread that should only have the JS code and the table.  We don't want to run any JS on the worker, that's for Rebol only...delegating to the GUI.  Running JavaScript on the worker is basically useless.

Sorry about that.  I'm looking into it right now.  But it seems you got a pretty broad picture of how this is put together!

**Comment by hostilefork** [comment]
Okay....I've addressed this in an inefficient, but hopefully illustrative way:

https://github.com/metaeducation/ren-c/commit/0cfe0526e44d1e7f5b1ed3b6b79efe5b5cb26cac

Works even with things like this one... :-)

     >> test: js-awaiter [i] { setTimeout(resolve, reb.UnboxInteger(reb.ArgR("i"))) }
     >> test 2000

Sorry if my remark misled you, but you definitely got the trial-by-fire of digging into the files like %make-reb-lib.r!

I'll get this refactored to do it more efficiently.  But I want to wait until I sort out the whole reject() and HALT stuff...

**Comment by BrianOtto** [comment]
Nice job! And great comments, it was very illustrative. I tested it locally, and all is working good :)

Thank you again for the quick fix! Now let's see what kind of UI framework we can make out of this ...

**Comment by hostilefork** [comment]
> Now let's see what kind of UI framework we can make out of this ...

Here's a little mash up that actually runs with DO now...  **do %webvid.reb**

https://github.com/hostilefork/replpad-js/commit/5b47969e8530f1559410b8b1de2bc23ae58741e5#diff-7720160aa52699d16ac64141696634a2R1

That is treating a FILE! like it should be loaded relative to the page root as the current directory (via fetch())

I've also added the ability to READ a GitHub blob URL.  We shouldn't go *too* far with any of this without some high-level organizational documents, but hard to know how to organize something if you haven't tried the basics...

**Comment by BrianOtto** [comment]
> Here's a little mash up that actually runs with DO now... **do %webvid.reb**

Awesome :) I have been working on something very similar for the last few days. I hope to have something up this weekend. It injects a CSS framework into the page and allows you to build out a UI using commands on the REPL. Still very experimental, but I wanted to test out some ideas I had, to see how they work in practice. Definitely will need to scope out how everything should be organized once we figure out what works and what doesn't.

**Comment by hostilefork** [comment]
@BrianOtto I see some commits on your branch and hate to interrupt your work!  But [I needed some peer review from a domain expert](https://stackoverflow.com/questions/55186667/detecting-returned-promise-only-status-of-an-async-function/55186684#comment97113123_55186667) on a technical problem, so I sync'd up the master.  There's some disruptions but hopefully the commits explain each.
 
Errors are coming along, slightly better--for now it's reporting them in the console log and it tries not to crash.  I'm still wrestling with what the design should be.

Overall several points of progress, though...!

**Comment by BrianOtto** [comment]
No problem. I'll attempt to sync up my branch tonight, so I don't get too far behind master.

Nice to hear on the progress :)

**Comment by hostilefork** [comment]
> No problem. I'll attempt to sync up my branch tonight, so I don't get too far behind master.

With some caveats, you should be able to move your demo out to a fully separate project now... READ of raw.github.com URLs works and can give you BINARY! or TEXT!, and DO is built on READ so the ability to read binary means that works.  I'm tinkering with GitLab now, should be able to read links from either...

Please do feel free to hang out in chat and ask about anything--your time is valuable (!) and if I can speed you up with answering questions in real-time all the better!
